cstore_metadata_serialization.c

```diff
@@ -579,3 +580,77 @@ ProtobufBinaryToDatum(ProtobufCBinaryData protobufBinary, bool datumTypeByValue,
 
 	return datum;
 }
+
+
+/* DeserializeFooterMetadata deserializes footer metadata to extract where footer
+ * starts and length of footer data in number of pages.
+ * Expects to see cstore magic number (char[12]), major version (uint32), minor version(uint32)
+ * startingPage (uint32), pageCount (uint32)
+ */
+bool
+DeserializeTableFooterMetadata(const char *data, uint32 length, uint32 *startingPage,
+							   uint32 *pageCount)
+{
+	uint32 offset = 0;
+	char *magicNumber = pstrdup(CSTORE_MAGIC_NUMBER);
+	uint32 magicLength = strnlen(magicNumber, NAMEDATALEN);
+	uint32 footerMetadataLength = magicLength + 4 * sizeof(uint32);
+	uint32 majorVersion = 0;
+	uint32 minorVersion = 0;
+
+	*startingPage = InvalidBuffer;
+	*pageCount = InvalidBuffer;
+
+	if (length < footerMetadataLength)
+	{
+		return false;
+	}
+
+	if (memcmp(data, magicNumber, magicLength) != 0)
+	{
+		return false;
+	}
+
+	offset = magicLength;
+
+	memcpy(&majorVersion, data + offset, sizeof(uint32));
+	offset += sizeof(uint32);
+
+	memcpy(&minorVersion, data + offset, sizeof(uint32));
+	offset += sizeof(uint32);
+
+	if (majorVersion != CSTORE_VERSION_MAJOR ||
+		minorVersion != CSTORE_VERSION_MINOR)
+	{
+		ereport(ERROR, (errmsg("Unsupported cstore_fdw version")));
```

Comment:
I'd expand this error to say that the relation has an unexpected cstore version. And also error messages start with a lower case...